### ISSUE TITLE
Cannot build on OSX 10.11.6 with CUDA 7.5 and XCode 7.3.1

### ISSUE THREAD

**Issue by grefab** [davisking/dlib#223]
Building dlib with CUDA support on OSX does not work (at least for me). Here is what I've tried:
- Building with clang from XCode 7.31 (Apple LLVM version 7.3.0 (clang-703.0.31) ): cudnn_test_build fails: Does not support thread_local (see http://stackoverflow.com/a/29929949/72082, my output here: https://gist.github.com/grefab/dfe4e64355d8bf13312ee446f7b346a4)
- Building with clang 3.8 from homewbrew: Build of cuda_test_build fails: That clang version is not supported by nvcc.
- Building with clang 3.7 from homebrew: cuda-test-build fails: Multiple definitions of pow (see https://gist.github.com/grefab/b6399975368d87bb8324748e1cc336dd) I'm not sure what to do about that.
- Building with GCC from homebrew (gcc-6 (Homebrew gcc 6.2.0) 6.2.0): cuda_test_build fails: nvcc fatal: GNU C/C++ compiler is no longer supported as a host compiler on Mac OS X.

Any ideas what I've missed and/or could be done to get dlib with CUDA support build on OSX?

**Comment by davisking** [comment]
Maybe you need clang 6 from x code (
http://docs.nvidia.com/cuda/cuda-getting-started-guide-for-mac-os-x/#axzz4JkoHKEy1).
In any case, to use dlib with cuda you need a compiler supported by NVIDIA
and the compiler needs to support C++11.

**Comment by grefab** [comment]
Meh. That won't help. C++11 support of Apple's current clang is incomplete, i.e. it does not support `thread_local`. Versions before that are no better. Compiler extensions (`__thread`, `_Thread_local`) are not the same (I tried replacing `thread_local` by them: `cudnn_dlibapi.cpp:166:13: error: '__thread' variables must have global storage`, `cudnn_dlibapi.cpp:115:13: error: '_Thread_local' variables must have global storage`, so unless I figure out a way to get rid of `thread_local` in my fork of dlib completely (which I really dont'want) it seems that I have to wait for XCode 8 with a version of clang that really supports C++11.

**Comment by davisking** [comment]
Well, you should be able to use the dlib::thread_specific_data instead.
You declare them statically in a function and it works basically the same
way as thread_local.

**Comment by davisking** [comment]
You need to declare the thread specific data static or the variable goes out of scope and you return a dangling reference.   So like this `static thread_specific_data<cublas_context> c;`.  Then it will probably work.

**Comment by grefab** [comment]
Whee! That did the trick! Thx! :)

**Comment by grefab** [comment]
Here is the source: https://github.com/grefab/dlib/commits/feature/osx_cuda_build This seems to be a more or less system-specific hack, so I'd keep it in a separate branch. Let me know if you want a pull request anyways.

**Comment by davisking** [comment]
Cool.  I'm not sure if I want to put this in dlib main since I find it somewhat absurd that `thread_local` is missing from clang. :\

If I did I would add some preprocessor logic that made `dlib::thread_specific_data` use `thread_local` when it can and it's current implementation when it can't.

**Comment by grefab** [comment]
Nah, I think what I've done is a workaround for a problem that should not be there in the first place (`thread_local` support), too. Xcode8 will be out the next days and should fix that. If not, I'll be happy to try to polish my solution accordingly. :)

**Comment by davisking** [comment]
Sounds good :)

**Comment by kylemcdonald** [comment]
the branch at https://github.com/grefab/dlib/commits/feature/osx_cuda_build mostly fixed my compilation issues.

at the moment, on OSX we have to use xcode 7.3 because CUDA doesn't work with 8.0. so even if 8.0 has a solution to `thread_local`, we can't use it yet. so the thing we're waiting on right now is for CUDA to issue a minor update for OSX users.
